# Atividade 2 — Exercícios de Limpeza, Transformação e Agrupamento de Dados

Responda as perguntas abaixo usando o dataset **emendas_tratamento.parquet**. o dataset pode ser baixado em https://drive.google.com/file/d/19ydA23-eboyiKxN29JL1EQWW4CplTiKg/view?usp=sharing.

Cada resposta deve ter:
- Uma célula de **código** com a solução
- Uma célula de **texto** (Markdown) com a sua interpretação do resultado

---
## Passo 0 — Importar pandas e carregar os dados

In [39]:
import pandas as pd 

dados = pd.read_parquet("https://drive.google.com/uc?export=download&id=19ydA23-eboyiKxN29JL1EQWW4CplTiKg")

#previa da tabela
display(dados.head())
dados.info()


,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
0,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,ALVORADA - RS,4300604,ALVORADA,4300000,...,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,700000.00,0.0,0.0,0.0,690900.0,9100.00,Sem informação
1,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,SANTA ISABEL DO PARÁ - PA,1506500,SANTA ISABEL DO PARÁ,1500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,487500.00,0.0,0.0,0.0,487500.0,0.00,Sem informação
2,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,BARRA BONITA - SC,4202099,BARRA BONITA,4200000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,100000.00,0.0,0.0,0.0,0.0,100000.00,Sem informação
3,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,ARARAQUARA - SP,3503208,ARARAQUARA,3500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,243750.00,0.0,0.0,0.0,243750.0,0.00,Sem informação
4,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,CAMPO LARGO - PR,4104204,CAMPO LARGO,4100000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,1658333.94,0.0,0.0,0.0,527200.0,1131133.94,Sem informação


<class 'pandas.DataFrame'>
RangeIndex: 88659 entries, 0 to 88658
Data columns (total 29 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Código da Emenda                    88659 non-null  str    
 1   Ano da Emenda                       88659 non-null  int64  
 2   Tipo de Emenda                      88659 non-null  str    
 3   Código do Autor da Emenda           72698 non-null  float64
 4   Nome do Autor da Emenda             72698 non-null  str    
 5   Número da emenda                    88659 non-null  str    
 6   Localidade de aplicação do recurso  88659 non-null  str    
 7   Código Município IBGE               88659 non-null  str    
 8   Município                           88659 non-null  str    
 9   Código UF IBGE                      88659 non-null  int64  
 10  UF                                  88659 non-null  str    
 11  Região                              88659 non-null  

## 1) Filtre apenas as emendas do ano de **2022** e mostre as 30 primeiras linhas. Em seguida responda: Quanta emendas foram registradas em 2022?

In [40]:
dados_2022 = dados[dados["Ano da Emenda"] == 2022]

display(dados_2022.head(30))
print(len(dados_2022))

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
63473,202238950001,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,1,CONCEIÇÃO DO ALMEIDA - BA,2908309,CONCEIÇÃO DO ALMEIDA,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,500000.00,0.00,0.00,500000.00,0.0,500000.00,202238950001
63474,202238950002,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,2,CONCEIÇÃO DO ALMEIDA - BA,2908309,CONCEIÇÃO DO ALMEIDA,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,750141.00,750141.00,750141.00,0.00,0.0,0.00,202238950002
63475,202238950003,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,3,CRUZ DAS ALMAS - BA,2909802,CRUZ DAS ALMAS,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,500000.00,0.00,0.00,500000.00,0.0,500000.00,202238950003
63476,202238950004,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,4,IRAMAIA - BA,2914307,IRAMAIA,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,200000.00,0.00,0.00,200000.00,0.0,200000.00,202238950004
63477,202238950005,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,5,MUCURI - BA,2922003,MUCURI,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,500000.00,0.00,0.00,500000.00,0.0,500000.00,202238950005
63478,202238950006,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,6,SERRA DO RAMALHO - BA,2930154,SERRA DO RAMALHO,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,535815.00,535815.00,535815.00,0.00,0.0,0.00,202238950006
63479,202238950007,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,7,SOUTO SOARES - BA,2930808,SOUTO SOARES,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,200000.00,0.00,0.00,200000.00,0.0,200000.00,202238950007
63480,202238950008,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,8,MÚLTIPLO,Sem informação,Múltiplo,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,470837.00,470837.00,470837.00,0.00,0.0,0.00,202238950008
63481,202238950009,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,9,MÚLTIPLO,Sem informação,Múltiplo,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,1100000.00,0.00,0.00,1100000.00,0.0,1100000.00,202238950009
63482,202238950010,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,10,MÚLTIPLO,Sem informação,Múltiplo,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,8573048.00,8123048.00,8123048.00,450000.00,0.0,450000.00,202238950010


6108


Foram registradas 6108 emendas no ano de 2022

## 2) Filtre as emendas da região **Norte** que tenham `Valor Empenhado` maior que **500.000** e mostre as 5 primeiras linhas. Em seguida, responda: quantas emendas atendem a esses duas condições ao mesmo tempo?

In [41]:
#vendo as regioes
display(dados["Região"].unique().tolist())

dados_norte_rico = dados[(dados["Região"] == "Norte") & (dados["Valor Empenhado"] > 500000)]

display(dados_norte_rico.head())
print(len(dados_norte_rico))


['Sul',
 'Norte',
 'Sudeste',
 'Nordeste',
 'Centro-Oeste',
 'Nacional',
 'Múltiplo',
 'Exterior']

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
18,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,MÂNCIO LIMA - AC,1200336,MÂNCIO LIMA,1200000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,616610.0,0.0,0.0,0.0,0.0,616610.0,Sem informação
120,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,VITÓRIA DO JARI - AP,1600808,VITÓRIA DO JARI,1600000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,522384.4,0.0,0.0,0.0,0.0,522384.4,Sem informação
160,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,SANTANA - AP,1600600,SANTANA,1600000,...,IMPLEMENTACAO DE INFRAESTRUTURA BASICA NOS MUN...,0,IMPLEMENTACAO DE INFRAESTRUTURA BASICA NOS MUN...,4443000.0,0.0,0.0,0.0,4443000.0,0.0,Sem informação
170,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,PALMEIRAS DO TOCANTINS - TO,1713809,PALMEIRAS DO TOCANTINS,1700000,...,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,935350.0,0.0,0.0,0.0,0.0,935350.0,Sem informação
218,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,COSTA MARQUES - RO,1100080,COSTA MARQUES,1100000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,549995.0,0.0,0.0,0.0,159996.0,389999.0,Sem informação


6230


Há 6230 emendas do norte com valor superior a 500 mil

## 3) Filtre as emendas que são do estado de **SÃO PAULO** ou do estado do **RIO DE JANEIRO** e mostre quantas linhas cada estado possui. Em seguida, responda: Qual dos dois estados possui mais emendas e quantas são?

In [42]:
#vendo os UF
display(dados["UF"].unique().tolist())

dados_sp_rj = dados[(dados["UF"] == "SÃO PAULO") | (dados["UF"] == "RIO DE JANEIRO")]

display(dados_sp_rj.head())
display(dados_sp_rj['UF'].value_counts())

['RIO GRANDE DO SUL',
 'PARÁ',
 'SANTA CATARINA',
 'SÃO PAULO',
 'PARANÁ',
 'PERNAMBUCO',
 'PARAÍBA',
 'GOIÁS',
 'BAHIA',
 'CEARÁ',
 'MINAS GERAIS',
 'MARANHÃO',
 'RIO GRANDE DO NORTE',
 'RIO DE JANEIRO',
 'ACRE',
 'MATO GROSSO DO SUL',
 'MATO GROSSO',
 'PIAUÍ',
 'ALAGOAS',
 'ESPÍRITO SANTO',
 'SERGIPE',
 'AMAZONAS',
 'RONDÔNIA',
 'AMAPÁ',
 'TOCANTINS',
 'DISTRITO FEDERAL',
 'RORAIMA',
 'Sem informação',
 'Múltiplo']

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
3,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,ARARAQUARA - SP,3503208,ARARAQUARA,3500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,243750.0,0.0,0.0,0.0,243750.0,0.0,Sem informação
16,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,ITALVA - RJ,3302056,ITALVA,3300000,...,APOIO A INFRAESTRUTURA PARA A EDUCACAO BASICA,0,INFRAESTRUTURA PARA A EDUCACAO BASICA - DESPES...,343000.0,0.0,0.0,0.0,343000.0,0.0,Sem informação
20,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,PILAR DO SUL - SP,3537909,PILAR DO SUL,3500000,...,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,245850.0,0.0,0.0,0.0,0.0,245850.0,Sem informação
30,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,VALENÇA - RJ,3306107,VALENÇA,3300000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,1399800.0,0.0,0.0,0.0,0.0,1399800.0,Sem informação
37,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,NaN,S/I,MONTE ALTO - SP,3531308,MONTE ALTO,3500000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,248757.0,0.0,0.0,0.0,0.0,248757.0,Sem informação


UF
SÃO PAULO         10435
RIO DE JANEIRO     6988
Name: count, dtype: int64

São Paulo tem mais emendas, 10435, que Rio de Janeiro, 6988

## 4) Na coluna `Código da Emenda`, substitua todos os valores `"Sem informação"` por `"Não disponível"`. Mostre a contagem de valores únicos da coluna antes e depois da substituição.

In [43]:
print(len(dados) - dados['Código da Emenda'].nunique())
dados = dados.replace({'Código da Emenda' : {'Sem informação': 'Não disponível'}})
print(len(dados) - dados['Código da Emenda'].nunique())

17928
17928


## 5) Na coluna `Nome Função`, transforme todos os valores para que fiquem em letras maiúsculas. Confirme a alteração mostrando os valores únicos da coluna.

In [44]:
dados['Nome Função'] = dados['Nome Função'].str.upper()
dados['Nome Função'].unique().tolist()

['URBANISMO',
 'COMÉRCIO E SERVIÇOS',
 'SAÚDE',
 'EDUCAÇÃO',
 'AGRICULTURA',
 'ASSISTÊNCIA SOCIAL',
 'DESPORTO E LAZER',
 'DEFESA NACIONAL',
 'SANEAMENTO',
 'DIREITOS DA CIDADANIA',
 'GESTÃO AMBIENTAL',
 'CULTURA',
 'CIÊNCIA E TECNOLOGIA',
 'SEGURANÇA PÚBLICA',
 'ORGANIZAÇÃO AGRÁRIA',
 'COMUNICAÇÕES',
 'ADMINISTRAÇÃO',
 'HABITAÇÃO',
 'INDÚSTRIA',
 'TRABALHO',
 'MÚLTIPLO',
 'TRANSPORTE',
 'PREVIDÊNCIA SOCIAL',
 'ENCARGOS ESPECIAIS',
 'RELAÇÕES EXTERIORES',
 'ENERGIA',
 'ESSENCIAL À JUSTIÇA']

## 6) Verifique quantas linhas duplicadas existem no dataset. Em seguida, remova as duplicatas e mostre quantas linhas sobraram.

In [45]:
#tamanho tabela e duplicadas
print(len(dados))
print(dados.duplicated().sum())

#removendo duplicadas
dados = dados.drop_duplicates()
print(len(dados))

88659
63
88596


## 7) Remova duplicatas considerando apenas as colunas `Código da Emenda` e `Município`. Quantas linhas sobraram? O resultado é diferente de quando removemos duplicatas do DataFrame inteiro? 

In [46]:
#tamanho da tabela e duplicadas
print(len(dados))
print(dados.duplicated(subset=['Código da Emenda', 'Município']).sum())

#tirando duplicados
dados = dados.drop_duplicates(subset=['Código da Emenda', 'Município'])
len(dados)

88596
12852


75744

Sobraram 75744 linhas, foram removidas 12852 enquanto foram removidas 63 duplicadas considerando o DataFrame todo

## 8) Identifique quais colunas possuem valores nulos e quantos nulos cada uma tem. Em seguida, responda: qual é a coluna que possui maior número de valores nulos e quantos são?

In [47]:
dados.isnull().sum()

Código da Emenda                         0
Ano da Emenda                            0
Tipo de Emenda                           0
Código do Autor da Emenda             4862
Nome do Autor da Emenda               4862
Número da emenda                         0
Localidade de aplicação do recurso       0
Código Município IBGE                    0
Município                                0
Código UF IBGE                           0
UF                                       0
Região                                   0
Código Função                            0
Nome Função                              0
Código Subfunção                         0
Nome Subfunção                           0
Código Programa                          0
Nome Programa                            0
Código Ação                              0
Nome Ação                                0
Código Plano Orçamentário                0
Nome Plano Orçamentário                  0
Valor Empenhado                          0
Valor Liqui

O Código e o Nome do Autor da Emenda tem 4862 valores nulos sendo as unicas colunas com valores nulos

## 9) Preencha os valores nulos da coluna `codigo do autor` com **-1** e os valores nulos da coluna `nome do autor da emenda` com "desconhecido" e confirme que não restaram nulos nessa coluna.

In [48]:
dados['Código do Autor da Emenda'] = dados['Código do Autor da Emenda'].fillna(-1)
dados['Nome do Autor da Emenda'] = dados['Nome do Autor da Emenda'].fillna('Desconhecido')

dados.isnull().sum()

Código da Emenda                      0
Ano da Emenda                         0
Tipo de Emenda                        0
Código do Autor da Emenda             0
Nome do Autor da Emenda               0
Número da emenda                      0
Localidade de aplicação do recurso    0
Código Município IBGE                 0
Município                             0
Código UF IBGE                        0
UF                                    0
Região                                0
Código Função                         0
Nome Função                           0
Código Subfunção                      0
Nome Subfunção                        0
Código Programa                       0
Nome Programa                         0
Código Ação                           0
Nome Ação                             0
Código Plano Orçamentário             0
Nome Plano Orçamentário               0
Valor Empenhado                       0
Valor Liquidado                       0
Valor Pago                            0


## 10) Verifique o tipo da coluna `Ano da Emenda`. Converta-a para **string** (`str`). Mostre o tipo antes e depois da conversão.

In [49]:
#verificando/convertendo tipo da coluna
print(dados['Ano da Emenda'].dtypes)
dados['Ano da Emenda'] = dados['Ano da Emenda'].astype(str)

print(dados['Ano da Emenda'].dtypes)


int64
str


## 11) Converta a coluna `Valor Empenhado` para o tipo **float** (caso ainda não seja). Mostre o tipo após a conversão.

In [50]:
#verificando tipo da coluna
print(dados['Valor Empenhado'].dtypes)

float64


Coluna ja esta em float

## 12) Calcule o **total** de `Valor Pago` por **Região**. Quais são as duas regiões que mais receberam recursos?

In [51]:
total_por_regiao = dados.groupby("Região")["Valor Pago"].sum().sort_values()

display(total_por_regiao)

Região
Exterior        2.218834e+07
Nacional        4.443190e+09
Centro-Oeste    9.445098e+09
Sul             1.488085e+10
Norte           1.811723e+10
Sudeste         2.982401e+10
Múltiplo        3.010563e+10
Nordeste        3.943671e+10
Name: Valor Pago, dtype: float64

Sudeste e Nordeste foram as regiões que mais receberam recursos com mais de 80 bilhões juntos

## 13) Calcule a **quantidade** de emendas por **Nome Função**. Mostre as 5 funções com mais emendas.

In [52]:
dados['Nome Função'].value_counts().head()

Nome Função
SAÚDE                 31637
EDUCAÇÃO               6796
URBANISMO              6627
ENCARGOS ESPECIAIS     4594
AGRICULTURA            3729
Name: count, dtype: int64

## 14) Crie uma nova coluna chamada `Saldo a Pagar`. Ela deve ser o resultado da subtração entre a coluna `Valor Empenhado` e a coluna `Valor Pago`. Mostre as 5 primeiras linhas apenas com essas três colunas.

In [53]:
#nova coluna saldo a pagar
dados['Saldo a Pagar'] = dados['Valor Empenhado'] - dados['Valor Pago']

dados[['Saldo a Pagar', 'Valor Pago', 'Valor Empenhado']].head()

,Saldo a Pagar,Valor Pago,Valor Empenhado
0,700000.00,0.0,700000.00
1,487500.00,0.0,487500.00
2,100000.00,0.0,100000.00
3,243750.00,0.0,243750.00
4,1658333.94,0.0,1658333.94


## 15) Elimine a(s) coluna(s) do dataframe que está totalmente vazia.

In [54]:
#verificando se as colunas são identicas
dados = dados.replace({'Código da Emenda_1' : {'Sem informação': 'Não disponível'}})
print(dados['Código da Emenda'].equals(dados['Código da Emenda_1']))

#removendo coluna
dados = dados.drop('Código da Emenda_1', axis=1)
print(dados.columns)

True
Index(['Código da Emenda', 'Ano da Emenda', 'Tipo de Emenda',
       'Código do Autor da Emenda', 'Nome do Autor da Emenda',
       'Número da emenda', 'Localidade de aplicação do recurso',
       'Código Município IBGE', 'Município', 'Código UF IBGE', 'UF', 'Região',
       'Código Função', 'Nome Função', 'Código Subfunção', 'Nome Subfunção',
       'Código Programa', 'Nome Programa', 'Código Ação', 'Nome Ação',
       'Código Plano Orçamentário', 'Nome Plano Orçamentário',
       'Valor Empenhado', 'Valor Liquidado', 'Valor Pago',
       'Valor Restos A Pagar Inscritos', 'Valor Restos A Pagar Cancelados',
       'Valor Restos A Pagar Pagos', 'Saldo a Pagar'],
      dtype='str')


## 16) Para finalizar, salve o DataFrame transformado em um arquivo no formato JSON.

In [55]:
#convertendo dataframe pra json
dados.to_json

display(dados)

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Saldo a Pagar
0,Não disponível,2014,Emenda Individual - Transferências com Finalid...,-1.0,Desconhecido,S/I,ALVORADA - RS,4300604,ALVORADA,4300000,...,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,700000.00,0.00,0.00,0.0,690900.0,9100.00,700000.00
1,Não disponível,2014,Emenda Individual - Transferências com Finalid...,-1.0,Desconhecido,S/I,SANTA ISABEL DO PARÁ - PA,1506500,SANTA ISABEL DO PARÁ,1500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,487500.00,0.00,0.00,0.0,487500.0,0.00,487500.00
2,Não disponível,2014,Emenda Individual - Transferências com Finalid...,-1.0,Desconhecido,S/I,BARRA BONITA - SC,4202099,BARRA BONITA,4200000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,100000.00,0.00,0.00,0.0,0.0,100000.00,100000.00
3,Não disponível,2014,Emenda Individual - Transferências com Finalid...,-1.0,Desconhecido,S/I,ARARAQUARA - SP,3503208,ARARAQUARA,3500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,243750.00,0.00,0.00,0.0,243750.0,0.00,243750.00
4,Não disponível,2014,Emenda Individual - Transferências com Finalid...,-1.0,Desconhecido,S/I,CAMPO LARGO - PR,4104204,CAMPO LARGO,4100000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,1658333.94,0.00,0.00,0.0,527200.0,1131133.94,1658333.94
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
88591,202544840005,2025,Emenda Individual - Transferências Especiais,4484.0,ZUCCO,5,MÚLTIPLO,Sem informação,Múltiplo,4300000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,12307991.08,12184912.08,12184912.08,0.0,0.0,0.00,123079.00
88592,202544840006,2025,Emenda Individual - Transferências com Finalid...,4484.0,ZUCCO,6,MÚLTIPLO,Sem informação,Múltiplo,4300000,...,FOMENTO AO SETOR AGROPECUARIO,0,FOMENTO AO SETOR AGROPECUARIO - DESPESAS DIVERSAS,988649.90,25592.00,25592.00,0.0,0.0,0.00,963057.90
88593,202544840007,2025,Emenda Individual - Transferências com Finalid...,4484.0,ZUCCO,7,Nacional,Sem informação,Sem informação,-1,...,ADMINISTRACAO DA UNIDADE,0,ADMINISTRACAO DA UNIDADE - DESPESAS DIVERSAS,999914.73,155246.57,149728.52,0.0,0.0,0.00,850186.21
88594,202544840008,2025,Emenda Individual - Transferências com Finalid...,4484.0,ZUCCO,8,Nacional,Sem informação,Sem informação,-1,...,"POLICIAMENTO, FISCALIZACAO, ENFRENTAMENTO A CR...",0,"POLICIAMENTO, FISCALIZACAO, ENFRENTAMENTO A CR...",497151.34,0.00,0.00,0.0,0.0,0.00,497151.34
